swift-sdk/Internal/InAppManager.swift

```diff
@@ -18,46 +26,121 @@ class InAppManager : IterableInAppManagerProtocol {
     }
     
     func show(message: IterableInAppMessage, consume: Bool = true, callback: ITEActionBlock? = nil) {
         ITBInfo()
         
-        // Handle url and call the client with callback provided
+        // This is public (via public protocol implementation), so make sure we call from Main Thread
+        DispatchQueue.main.async {
+            _ = self.showInternal(message: message, consume: consume, checkNext: false, callback: callback)
+        }
+    }
+
+    func remove(message: IterableInAppMessage) {
+        ITBInfo()
+
+        message.consumed = true
+        internalApi?.inAppConsume(message.messageId)
+        
+        messagesMap.updateValue(message, forKey: message.messageId)
+    }
+
+    @objc func onAppEnteredForeground(notification: Notification) {
+        ITBInfo()
+        processMessages()
+    }
+    
+    // This must be called from MainThread
+    private func showInternal(message: IterableInAppMessage,
+                              consume: Bool,
+                              checkNext: Bool,
+                              callback: ITEActionBlock? = nil) -> Bool {
+        ITBInfo()
+
+        guard Thread.isMainThread else {
+            ITBError("This must be called from the main thread")
+            return false
+        }
+        
+        // This is called when the user clicks on a link in the inAPP
         let clickCallback = {(urlOrAction: String?) in
             // call the client callback, if present
             callback?(urlOrAction)
+            
             // in addition perform action or url delegate task
             if let urlOrAction = urlOrAction {
                 self.handleUrlOrAction(urlOrAction: urlOrAction)
             } else {
                 ITBError("No name for clicked button/link in inApp")
             }
+            
+            // check if we need to check for more inApps after showing this
+            // This will be true when we are processing messagers from server and false
+            // when called by clients directly
+            if checkNext == true {
+                self.queue.asyncAfter(deadline: .now() + self.retryInterval) {
+                    self.processMessages()
+                }
+            }
         }
         
-        displayer.showInApp(message: message, callback: clickCallback).onSuccess { (showed) in // showed boolean value gets set when inApp is showed in UI
-            if showed && consume {
-                self.internalApi?.inAppConsume(message.messageId)
-                self.messagesMap.removeValue(forKey: message.messageId)
-            } else {
-                message.skipped = true
-                self.messagesMap.updateValue(message, forKey: message.messageId)
-            }
+        // set processed, if not already set
+        message.processed = true
+
+        let showed = displayer.showInApp(message: message, callback: clickCallback)
+        if showed {
+            lastShowedTime = dateProvider.currentDate
```

Comment:
`lastShowedTime` should be set when the in-app is dismissed, so we count the interval as the time between in-app dismiss and next in-app show.